### PR TITLE
HPCC-25219 add placement configuration support.

### PR DESCRIPTION
**Pull request by xwang2713**
This PR replaces https://github.com/hpcc-systems/HPCC-Platform/pull/14582
The configuration includes followings per pod,target or type:
nodeSelector,tolerations,affinity and schedulerName.

\#\# placements
\#\# The placement is responsible for finding the best Node for a Pod. It can be configured
\#\# through placement to include an array of objects to configure the Kubernetes Scheduler.
\#\# It must have a "pods" list which tells it which pod the settings will be applied to.
\#\# The synatx is:
\#\#   placements:
\#\#   - pods [list]
\#\#     placement:
\#\#       <supported configurations>
\#\#
\#\# The list item in "pods" can be one of the following:
\#\# 1) HPCC Systems component types in format: "type:<type name>". <type name> includes
\#\#    dali, esp, eclagent, eclccserver, roxie, thor. For example "type:esp"
\#\# 2) Target, the name of array item of above type in format "target:<target name>"
\#\#    For example "target:roxie", "target:thor".
\#\# 3) Pod, "Deployment" metadata name, which usually should be from the name of the array
\#\#    item of a type. For example, "eclwatch", "mydali", "thor-thoragent"..
\#\# 4) Job name regular express:  For example "compile-" or "compile-.*" or exact match "^compile-.*$"
\#\#
\#\# Supported configurations under each "placement"
\#\# 1) nodeSelector
\#\# 2) taints/tolerations
\#\# 3) affinity
\#\#    There is no schema check for the content of affinity. Reference
\#\#    https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/
\#\# 4) schedulerName: profile names. "affinity" defined in scheduler profile  requires
\#\#    Kubernetes 1.20.0 beta and later releases
\#\#
\#\# Examples:
\#placements:
\#- pods: ["eclwatch","roxie-workunit","^compile-.*$","mydali"]
\#  placement:
\#    nodeSelector:
\#      name: npone
\#- pods: ["thorworker-", "thor-thoragent", "thormanager-","thor-eclagent","hthor-"]
\#  placement:
\#    nodeSelector:
\#      name: nptwo
\#    tolerations:
\#    - key: "name"
\#      operator: "Equal"
\#      value: "nptwo"
\#      effect: "NoSchedule"
\#- pods: ["thorworker-.*"]
\#  placement:
\#    affinity:
\#      nodeAffinity:
\#        requiredDuringSchedulingIgnoredDuringExecution:
\#          nodeSelectorTerms:
\#          - matchExpressions:
\#            - key: kubernetes.io/e2e-az-name
\#              operator: In
\#              values:
\#              - e2e-az1
\#              - e2e-az2
\#- pods: ["target:roxie"]
\#  placement:
\#    schedulerName: "my-scheduler"
\#\# The settings will be applied to all thor pods/jobs and myeclccserver pod and job
\#- pods: ["target:myeclccserver", "type:thor"]
\#  placement:
\#    nodeSelector:
\#      app: "tf-gpu"
\#    tolerations:
\#    - key: "gpu"
\#      operator: "Equal"
\#      value: "true"
\#      effect: "NoSchedule"

Signed-off-by: Xiaoming Wang xiaoming.wang@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

1) uncomment "placements" examples in values.yaml.
   Use "helm template" to generate yaml content to a file.
   Check the contents
2) Only uncomment "nodeSelector" and "tolerations" in "placement" of the values.yaml.
    Start AKS by running start.sh in https://github.com/gfortil/HPCC-Tests/tree/main/Azure/Kubernetes/Scheduler-Test/aks-np-1
    Deploy HPCC cluster 
    Validate the node assigning according the nodeSelector and tolerations settings
    Delete AKS cluster by running delete-resource-group.sh in https://github.com/gfortil/HPCC-Tests/tree/main/Azure/Kubernetes/Scheduler-Test/aks-np-1

### CODE REVIEW

File: helm/hpcc/values.yaml [hunk 1]
```diff
@@ -49,6 +49,75 @@ security:
     extern: "allow"
     datafile: "allow"
 
+## placements
+## The placement is responsible for finding the best Node for a Pod. It can be configured
+## through placement to include an array of objects to configure the Kubernetes Scheduler.
+## It must have a "pods" list which tells it which pod the settings will be applied to.
+## The synatx is:
```
Comment (richardkchapman) [inline helm/hpcc/values.yaml:56]: Typo

File: helm/hpcc/values.yaml [hunk 2]
```diff
@@ -49,6 +49,75 @@ security:
     extern: "allow"
     datafile: "allow"
 
+## placements
```
Comment (richardkchapman) [inline helm/hpcc/values.yaml:52]: As a general comment (not specific to this PR but we need to think about it), the values.yaml is starting to get very large. Do we want so much in-line documentation? Is this the best location for the information, and is it starting to make the values.yaml hard to read/follow? Is there a better solution?

**Review by richardkchapman — CHANGES_REQUESTED** [review]

**Comment by jakesmith** [comment]
@xwang2713 - something may have merged since you pushed, but is now conflicting. Can you rebase ?

File: helm/hpcc/values.yaml (hunk 2 above)
Comment (xwang2713) [reply helm/hpcc/values.yaml:52]: Another solution is to instruct user:
1) helm template to generate yaml file
2) use kustomize or similar tool to patch whatever changes 
3) kubectl apply/create to deploy the cluster
4) kubectl patch to change dynamic settings

File: helm/hpcc/values.yaml (hunk 1 above)
Comment (xwang2713) [reply helm/hpcc/values.yaml:56]: Fixed

**Comment by xwang2713** [comment]
Fixed typo and rebase with master

File: helm/hpcc/values.yaml (hunk 2 above)
Comment (richardkchapman) [reply helm/hpcc/values.yaml:61]: I don't object to the values file being the place where we set this information.

I'm questioning whether the default values file is the best place to document this information

**Comment by jakesmith** [comment]
@xwang2713 - still appears to be basing problem (see 9 commits).

**Comment by xwang2713** [comment]
Rebase again as @jakesmith suggested

**Comment by xwang2713** [comment]
Wait. Still have rebase problem

File: helm/hpcc/values.yaml (hunk 2 above)
Comment (xwang2713) [reply helm/hpcc/values.yaml:63]: Maybe just point a document for user and move all instructions including examples to the documentation?. I am OK with it.

Comment (jakesmith) [reply helm/hpcc/values.yaml:63]: I think that's sensible, but there are pre-existing sections that could also benefit from moving out into a document.
Let's leave this as it is for now, and tackle tidying up/referencing doc(s) from values.yaml in a separate JIRA/PR

Comment (xwang2713) [reply helm/hpcc/values.yaml:63]: Or put placements instructions to a file, such as Placements.md

**Comment by xwang2713** [comment]
Rebased.
Move instruction and examples to file docs/placements.md

File: helm/hpcc/templates/eclccserver.yaml [hunk 3]
```diff
@@ -3,6 +3,7 @@ EclccServer configmap
 Pass in dict with root and me
 */}}
 {{- define "hpcc.eclccServerConfigMap" -}}
+{{- $compileJobName := printf "compile-%%jobname" }}
```
Comment (jakesmith) [inline helm/hpcc/templates/eclccserver.yaml:6]: This change doesn't look necessary, the printf isn't acutally doing anything here (no arguments).
Could revert this and line 27 to way it was.

**Review by jakesmith — APPROVED** [review]
@xwang2713 - looks good to me.

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**